### ISSUE TITLE
Outstanding Primitive Rules for Higher Order Automatic Differentiation (jet)

### ISSUE THREAD

**Issue by jessebett** [google/jax#2431]
- [ ] zeros_like
- [ ] id
- [ ] add_any
- [x] device_put
- [x] neg
- [ ] sign
- [ ] floor
- [ ] ceil
- [ ] round
- [ ] is_finite
- [x] exp
- [x] log
- [ ] expm1
- [ ] log1p
- [ ] tanh
- [ ] sin
- [ ] cos
- [ ] atan2
- [ ] regularized_incomplete_beta
- [ ] lgamma
- [ ] igamma
- [ ] igammac
- [ ] bessel_i0e
- [ ] bessel_i1e
- [ ] erf
- [ ] erfc
- [ ] erf_inv
- [x] real
- [ ] imag
- [x] complex
- [ ] conj
- [ ] abs
- [ ] sqrt
- [ ] rsqrt
- [ ] pow
- [ ] and
- [ ] or
- [ ] xor
- [x] add
- [x] sub
- [x] mul
- [ ] safe_mul
- [x] div
- [ ] rem
- [ ] max
- [ ] min
- [ ] shift_left
- [ ] shift_right_arithmetic
- [ ] shift_right_logical
- [ ] eq
- [ ] ne
- [ ] ge
- [ ] gt
- [ ] le
- [ ] lt
- [x] convert_element_type
- [ ] bitcast_convert_type
- [x] conv_general_dilated
- [x] dot_general
- [x] broadcast
- [x] broadcast_in_dim
- [ ] clamp
- [x] concatenate
- [x] pad
- [x] reshape
- [x] rev
- [x] transpose
- [ ] select
- [x] slice
- [ ] dynamic_slice
- [ ] dynamic_update_slice
- [x] gather
- [ ] scatter-add
- [ ] scatter
- [x] reduce_sum
- [ ] reduce_prod
- [x] reduce_max
- [ ] reduce_min
- [x] reduce_window_sum
- [ ] reduce_window_max
- [ ] reduce_window_min
- [ ] select_and_scatter_add
- [ ] select_and_gather_add
- [ ] sort
- [ ] sort_key_val
- [x] tie_in
- [ ] stop_gradient
- [ ] while
- [ ] cond
- [ ] scan
- [ ] root
- [ ] custom_linear_solve
- [ ] fft
- [ ] psum
- [ ] ppermute
- [ ] jit(arcsinh,static_argnums=())
- [ ] cholesky
- [ ] eigh
- [ ] triangular_solve
- [ ] lu
- [ ] qr
- [ ] svd
- [ ] jit(slogdet, static_argnums=())
- [ ] logit
- [ ] expit
- [ ] log_ndtr
- [ ] random_gamma

@jacobjinkelly @duvenaud @mattjj

Here's the list of all the `ad.jvp_primitives` and probably reflects what we'd need for coverage by `jet`.

Copied from #2363 since that was more about merging into master!

**Comment by jessebett** [comment]
To anyone who wants to follow along, (or contribute!) these rules all found at the bottom of [`jax/experimental/jet.py`](https://github.com/google/jax/blob/master/jax/experimental/jet.py#L127-L244)

**Comment by jacobjinkelly** [comment]
If we want to be comprehensive should we also include `sinh`, `cosh`, `asinh`, `acosh`, `digamma`?

**Comment by jacobjinkelly** [comment]
I think `digamma`, `igamma`, `igammac`, and `lgamma` are all impossible due to [Hölder's theorem](https://en.wikipedia.org/wiki/H%C3%B6lder%27s_theorem).

**Comment by duvenaud** [comment]
Wow, great point Jacob.  I suppose there's the slim possibility that the ODE coefficients would be expressible in terms of the gamma function itself?

**Comment by jacobjinkelly** [comment]
Yeah I think that's possible! But I think such a scenario wouldn't characterize the gamma function, see [here](https://en.wikipedia.org/wiki/Gamma_function#19th%E2%80%9320th_centuries:_characterizing_the_gamma_function).

**Comment by shoyer** [comment]
If there's interest in defining jet rules for numerical linear algebra, I would suggest referencing [AlgoPy](https://github.com/b45ch1/algopy) and  [Sebastian Walter's PhD thesis](https://pdfs.semanticscholar.org/e52d/834dac4fdfdc4328105e29ae657c3fa51467.pdf).

AlgoPy has [higher order auto-diff implementations](https://github.com/b45ch1/algopy/blob/master/algopy/utpm/utpm.py) for the full range of JAX's linear algebra primitives: `linear_solve` (which covers both `custom_linear_solve` and `triangular_solve`), `cholesky`, `lu`, `qr`, `eigh` and `svd`.

**Comment by jessebett** [comment]
@shoyer this is an awesome reference! Thanks for making me aware of it I'm looking through it now and agree that these are great sources for those linalg primitives!

**Comment by pnkraemer** [comment]
Hi all!

Sorry to jump in from the side on this discussion. I am a big fan of `jet`, and some more primitives would be super useful to me! 
Are there still plans on adding some of the missing pieces in the future?

For example, `jnp.roll` requires the `dynamic_slice` rule, which stops me from using `jet` on the Lorenz96 ODE problem (there are workarounds for Lorenz96, but I found the `roll` implementation to be most efficient/readable).

### Example:

```python
import jax.numpy as jnp
from jax.experimental.jet import jet 

def f_lorenz96(y, c=1.):
    A = jnp.roll(y, shift=-1)
    B = jnp.roll(y, shift=2)
    C = jnp.roll(y, shift=1)
    D = y
    return (A - B) * C - D + c

y0 = jnp.ones(11) # values dont matter

jet(f_lorenz96, (y0,), ((y0,),)) 
```  
which raises a `KeyError`
```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "example.py", line 16, in <module>
    jet(f_lorenz96, (y0,), ((y0,),))
  File "/home/kraemer/Projects/high-dimensional-ode-solver/sth/env/lib/python3.8/site-packages/jax/experimental/jet.py", line 57, in jet
    out_primals, out_terms = jet_fun(jet_subtrace(f), order).call_wrapped(primals, series)
  File "/home/kraemer/Projects/high-dimensional-ode-solver/sth/env/lib/python3.8/site-packages/jax/linear_util.py", line 166, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "example.py", line 6, in f_lorenz96
    A = jnp.roll(y, shift=-1)
  File "/home/kraemer/Projects/high-dimensional-ode-solver/sth/env/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 5409, in roll
    return _roll(a, shift, axis)
  File "/home/kraemer/Projects/high-dimensional-ode-solver/sth/env/lib/python3.8/site-packages/jax/experimental/jet.py", line 140, in process_call
    result = call_primitive.bind(f_jet, *primals_and_series, **new_params)
  File "/home/kraemer/Projects/high-dimensional-ode-solver/sth/env/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 5386, in _roll
    return lax.reshape(_roll(ravel(a), shift, axis=0), a_shape)
  File "/home/kraemer/Projects/high-dimensional-ode-solver/sth/env/lib/python3.8/site-packages/jax/experimental/jet.py", line 140, in process_call
    result = call_primitive.bind(f_jet, *primals_and_series, **new_params)
  File "/home/kraemer/Projects/high-dimensional-ode-solver/sth/env/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 5400, in _roll
    a = lax.dynamic_slice_in_dim(a, a_shape[i] - x, a_shape[i], axis=i)
  File "/home/kraemer/Projects/high-dimensional-ode-solver/sth/env/lib/python3.8/site-packages/jax/experimental/jet.py", line 126, in process_primitive
    rule = jet_rules[primitive]
KeyError: dynamic_slice
```
I could make a PR with a little guidance (for example, if pointed to the correct primitive).
Also, please let me know if this is too off-topic for this discussion and I should open a new issue instead. :) 

Thanks a lot!


#### In case it is necessary: 
Python: 3.8.10
Pip freeze:
```
absl-py==1.0.0
flatbuffers==2.0
jax==0.2.28
jaxlib==0.1.76
numpy==1.22.2
opt-einsum==3.3.0
scipy==1.7.3
six==1.16.0
typing-extensions==4.0.1
```

**Comment by duvenaud** [comment]
That's great that you're using jets!  I can help you debug.  I'm not sure this will work, but my guess is that, like `slice`, `dynamic_slice` is linear.  In those cases, the jet rule is to simply apply the same operation to all the elements of the series.  We made a generic helper for such functions called `def_linear`.

Here it is applied to slice:
https://github.com/google/jax/blob/main/jax/experimental/jet.py#L240

So I think you might just be able to add a line for `dynamic_slice` everywhere you see a line for `slice`.

**Comment by pnkraemer** [comment]
Great, thanks a lot for the quick reply! I will give this a try. 

Would you like me to add a test (e.g. to https://github.com/google/jax/blob/7a6986c4c8fd8469bae36306efb0417b0a2f6d8c/tests/jet_test.py#L306) and make a pull request?

**Comment by duvenaud** [comment]
Yes, please!

**Comment by pnkraemer** [comment]
I think `dynamic_slice` can now be checked off in the list above. I cannot do it myself, which is why I comment.

If you need more help, please let me know. Otherwise I will come back next time I need a jet primitive that has not been implemented yet :) 

Thanks for the support!

**Comment by pnkraemer** [comment]
I've had another look, and many of the primitives above are already implemented. It probably makes sense to check them off in the todo list above. :)

More specifically (in the order of appearance):

* is_finite: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L210
* expm1: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L296
* log1p: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L297
* sin: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L486
* cos: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L487
* atan2: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L455
* erf: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L286
* erfc: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L303
* erf_inv: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L348
* imag: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L229
* conj: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L228
* pow: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L384
* rem: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L304
* shift_left: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L211
* shift_right_arithmetic: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L212
* shift_right_logical: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L213
* bitcast_convert_type: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L214
* clamp: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L305
* scatter-add: https://github.com/google/jax/blob/c49fb9c2808c0c1cca78a78a0f998bf25fa2fb42/jax/experimental/jet.py#L603

**Comment by jacobjinkelly** [comment]
@pnkraemer Thanks for going through and making this list! I can confirm that the jet rule for `expit` is implemented and works based on these tests:
https://github.com/google/jax/blob/main/tests/jet_test.py#L278-L281

**Comment by dsheldon** [comment]
Have you considered a generic polynomial composition routine to propagate Taylor series? It could be used for g(f(x)) whenever the higher derivatives of g(z) with respect to z are available and fill in cases where specialized routines aren't available. I implemented a basic one that runs in O(K^3) time and tested it for `lgamma` and `digamma`:

```python
def poly_mul(f, g):
  '''Multiply polynomials, truncating to larger degree'''
  f, g = (f, g) if len(f) >= len(g) else (g, f)  # ensure f has larger degree
  k = len(f)
  out = [jnp.zeros_like(f[0]) for i in range(k)]
  # naive convolution
  for i in range(k):
    g_lim = np.minimum(len(g), k-i)
    for j in range(g_lim):
      out[i+j] += f[i]*g[j]
  return out

def poly_compose(g, f):
  '''Polynomial composition via Horner's method (recursive)'''
  if not(g):
    return [jnp.zeros_like(f[0]) for term in f]
  res = poly_mul(f, poly_compose(g[1:], f))
  return [g[0] + res[0]] + res[1:]

def _series_to_taylor(f):
  # Add zero term and scale with factorials
  return [jnp.zeros_like(f[0])] + [term/fact(i+1) for i, term in enumerate(f)]

def _taylor_to_series(f):
  # Unscale and remove constant term
  return [term*fact(i+1) for i, term in enumerate(f[1:])]
  
def _compose_taylor(g, f):
  # g and f are sequences of derivatives
  # convert to Taylor polynomials and compose
  g = _series_to_taylor(g)
  f = _series_to_taylor(f)
  h = poly_compose(g, f)
  return(_taylor_to_series(h))

def _faa_di_bruno_rule(fun, local_deriv_fun, primals_in, series_in):
  x, = primals_in
  series, = series_in
  primals_out = fun(x)
  series_local = local_deriv_fun(x, len(series))
  series_out = _compose_taylor(series_local, series) if series else []
  return primals_out, series_out

_lgamma_local_derivs = lambda x, k: [jax.scipy.special.polygamma(n, x) for n in range(k)]
jet_rules[lax.lgamma_p] = partial(_faa_di_bruno_rule, lax.lgamma, _lgamma_local_derivs)

_digamma_local_derivs = lambda x, k: [jax.scipy.special.polygamma(n, x) for n in range(1, k+1)]
jet_rules[lax.digamma_p] = partial(_faa_di_bruno_rule, lax.digamma, _digamma_local_derivs)
```

The implementation could probably be optimized. Also, for large K asymptotically faster algorithms for composition (e.g., [Brent and Kung, 1978](https://dl.acm.org/doi/10.1145/322092.322099)) might be beneficial.

**Comment by jacobjinkelly** [comment]
Hi @dsheldon thanks for your comment, this is awesome :) Really clean implementation and nice examples!

Feel free to make a PR for this :D 

I believe @dougalm, @duvenaud, @jessebett, @mattjj worked on an implementation of FDB with some optimisations [here](https://github.com/jessebett/jax/blob/jet/jax/interpreters/fdb.py), but we never managed to merge this into the jet API. So feel free to have a peak at that code if you're keen and incorporate some of those ideas into your PR too :) 

IIRC @jessebett implemented some ideas from Brent and Kung, but I can't find the code for that unfortunately. @jessebett would you be able to provide a pointer?

**Comment by dsheldon** [comment]
Hi @jacobjinkelly, thanks for the response! Sure, I'd be happy to make a PR. Before I do: did I make the correct assumption about the inputs to the propagation rules? I.e., that `series` in the `_faa_di_bruno_rule` above will always be a list of arrays (or array-like objects) with the same shape?

I think the other implementation you pointed to is iterating through integer partitions, which, to my understanding, would take super-polynomial time. With polynomial composition it's possible to implement Faa di Bruno's rule in polynomial time — e.g., $O(k^3)$ for the naive algorithm above. One of the reasons for my comment was to point out this possibility.

We used one of the Brent-Kung algorithms in [a paper](https://arxiv.org/abs/1806.03207), so I have some old code, but I'm guessing the crossover point where it will be substantially faster than Horner's rule doesn't occur until the order of differentiation is pretty large. (And for very large $k$ it's hard to get around numerical issues due to dealing with terms like $k!$ in floating point anyway.)

**Comment by jacobjinkelly** [comment]
Yes I think that's the correct assumption for the type of `series`. If you're not sure, you can always double check by using the `unary_check` utility in [jet_test.py](https://github.com/google/jax/blob/main/tests/jet_test.py#L300).

> I think the other implementation you pointed to is iterating through integer partitions, which, to my understanding, would take super-polynomial time. With polynomial composition ...

Ok wow I didn't realise that, that's super interesting! 

> substantially faster than Horner's rule doesn't occur until the order of differentiation is pretty large ...

Agreed!

**Comment by jessebett** [comment]
Hi @dsheldon, 

As Jacob already said, this is a very nice implementation!

It's been some time so I'm a bit fuzzy on the details, sorry in advance if I'm missing your main points.

>Have you considered a generic polynomial composition routine to propagate Taylor series?

Yes! Early implementations we worked with exclusively relied on Faa Di Bruno and generic polynomial propagation. We included some notes on the relationship between Faa Di Bruno formula and propagation of truncated Taylor polynomials in the appendices of our [workshop preprint](https://openreview.net/pdf?id=SkxEF3FNPH). 

As you noted, Brent and Kung (1978) suggests asymptotically faster algorithms for the general polynomial composition (following from algorithmic complexity for inverting power series). However, as described in Griewank and Walther (2008) Chapter 13.2 on _Taylor Polynomial Propagation_, specifically the subsection on _Nonlinear Univariate Elementals_, generic polynomial composition addresses the general case where the Taylor coefficients are unrelated. But in practice the coefficients for "elemental functions of interest" (e.g. primitives like `exp` and `sin`) all are solutions of linear ODEs and satisfy the identity in equation 13.7. This allows application of Brent and Kung Theorem 5.1 to derive more efficient algorithms by exploiting relationships between coefficients. 

So at some point we moved from generic Faa Di Bruno based propagation rules to more optimized rules for primitives that are a consequence of this special relationship among their coefficients. You can find some examples of these rules and their complexity in tables 13.1 and 13.2 of Griewank and Walther. 

To be clear, Brent and Kung discuss algorithms for efficient propagation of power series with better complexity than classical polynomial propagation, but those improvements are really only relevant at larger orders of differentiation. You already point out this is numerically tricky due to $k!$. Instead, we are appealing to a restriction from the general problem of polynomial composition to those with special structure, which is a separate result also discussed in Brent and Kung.


> did I make the correct assumption about the inputs to the propagation rules? I.e., that series in the` _faa_di_bruno_rule` above will always be a list of arrays (or array-like objects) with the same shape?

Yes you got it. To help orient through the current code, you can see/modify the enforcement of the series input type [here](https://github.com/google/jax/blob/71b896893aefab32c11702ca99c8e7865f384700/jax/experimental/jet.py#L181)
and the enforcement that the series have the same shape [here](https://github.com/google/jax/blob/71b896893aefab32c11702ca99c8e7865f384700/jax/experimental/jet.py#L125-L129).


> I think the other implementation you pointed to is iterating through integer partitions, which, to my understanding, would take super-polynomial time.

Yes, Jacob linked to an implementation preceding our use of Faa Di Bruno rules that relies on integer partitions. Your understanding is correct, this was very non-optimal! A subsequent version that uses Faa Di Bruno exclusively is somewhere in the commit/branch history, but I'd have to do some digging to find it...

> It could be used for `g(f(x))` whenever the higher derivatives of `g(z)` with respect to `z` are available and fill in cases where specialized routines aren't available.

I agree, this would be great to have! At some point we discussed keeping the generic Faa Di Bruno rule as a fallback for primitives without special rules. I'm not quite sure what happened that dropped the generic rule, but I suspect we just got distracted by working out special rules for primitives we encountered in practice and the general case became increasingly rare. As Jacob said, if you're up for it a PR would be very welcome!

Thanks again for your comment and clean code!

**Comment by dsheldon** [comment]
OK, PR created --- hopefully done correctly, if not please let me know!

@jessebett, thanks for your comments, I think we're on the same page regarding the options! The optimized rules (usually O(k^2)) when the function is the solution of an ODE are preferable and applicable in almost all cases. I made my comment because the [notes explaining jet](https://github.com/google/jax/files/6717197/jet.pdf) mention (1) the intractable direct evaluation of Faa Di Bruno's formula, and (2) the efficient special case for common primitives that are the solution of a linear ODE —but they don't mention that a generic rule based on polynomial composition runs in O(k^3) (or better time), which is slower but not disastrously slower than the specialized rules. So it seemed like such a rule could be helpful for the few primitives that weren't already covered by special cases. From your response, it sounds like you already had this in previous versions.

**Comment by mattjj** [comment]
@dsheldon thanks for the tips!

I remember trying polynomial composition, but I thought it didn't give the correct answers (even though I thought it should work), and I never figured out why! I think I messed up `_series_to_taylor` and `_taylor_to_series` (referring to the names in your PR). It's great if this works generically!

**Comment by yachmena** [comment]
We would like to use jet to build Taylor series expansions of the rovibrational kinetic energy operator, employed for computing rovibrational spectra. The implementation involves differentiating through matrix inversion, which is currently not supported. Are there any plans to introduce custom_linear_solve primitive rules in the near future?

**Comment by shoyer** [comment]
No, I think it's safe to assume that `jet` is not being actively maintained.

**Comment by mattjj** [comment]
Actually that’s not right: jet is used pretty extensively, by projects like a [recent NeurIPS best paper](https://openreview.net/forum?id=J2wI2rCG2u), some other research projects in ML and scientific computing, and some GDM projects, and as a result we are happy to fix and improve it. That includes supporting new primitives as the need crops up.

Can you open a feature request issue? Or maybe two: one for the matrix ops you need, and another for a general custom rule setup? Tag me when you do!

**Comment by yachmena** [comment]
Thank you @mattjj, will follow your suggesion!

In the meantime, I’ve come up with a solution that seems to produce the correct results (for matrix inversion), though it is not the most appropriate approach. Based on the examples in jet, it looks like I just need to define differentiation rules for lax functions. For instance, I can register a rule using:
```python
jet.jet_rules[lax.linear_solve_p] = partial(_inverse_taylor_rule, lax.linear_solve_p)
```

However, I’m still unsure about how lax.custom_linear_solve is handled in JAX. Instead of using jax.linalg.inv, I registered a new function inv and defined Taylor series rules for it as follows:
```python

def fact(n):
    return lax.exp(lax.lgamma(n + 1.0))


inv_p = Primitive("inv")


def inv(a, **kw):
    return inv_p.bind(a, **kw)


def inv_impl(a, **kw):
    return jnp.linalg.inv(a)


inv_p.def_impl(inv_impl)


def inv_jvp(primals, tangents):
    a = primals
    da = tangents
    a_inv = jnp.linalg.inv(a)
    da_inv = -jnp.dot(a_inv, jnp.dot(da, a_inv))
    return a_inv, da_inv


ad.primitive_jvps[inv_p] = inv_jvp


def inv_abstact_eval(a, **kw):
    return ShapedArray(a.shape, a.dtype)


inv_p.def_abstract_eval(inv_abstact_eval)


def inv_lowering(ctx, a, **kw):
    return mlir.lower_fun(jnp.linalg.inv, multiple_results=False)(ctx, a)


mlir.register_lowering(inv_p, inv_lowering)


@jax.jit
def _inverse_taylor_rule(primals_in, series_in, **kw):
    (x,) = primals_in
    (x_terms,) = series_in
    u = [x] + x_terms
    v = [None] * len(u)
    if 'inv0' in kw: # to avoid repetitative inversion for different sets of series coefficients
        xi = kw['inv0']
    else:
        xi = jnp.linalg.inv(x)
    v[0] = xi

    def scale(k, j):
        return 1.0 / (fact(k - j) * fact(j))

    for k in range(1, len(v)):
        v[k] = (
            -fact(k) * xi @ sum(scale(k, j) * u[j] @ v[k - j] for j in range(1, k + 1))
        )
    primal_out, *series_out = v
    return primal_out, series_out

jet.jet_rules[inv_p] = _inverse_taylor_rule
```
